syntax = "proto3";

enum BlockType {
  EMPTY = 0;
  LEFT_UP = 1;  // ╝
  LEFT_DOWN = 2; // ╗
  RIGHT_DOWN = 3; // ╔
  RIGHT_UP = 4; // ╚
  LEFT_RIGHT = 5; // ═
  UP_DOWN = 6; // ║
}

message Block {
  BlockType block_type = 1;  // Use enum instead of string for compact representation
}

message Row {
  repeated Block blocks = 1;  // Each row contains multiple blocks
  uint32 row_clue = 2;        // Clue at the end of the row
}

message Grid {
  repeated Row rows = 1;  // The puzzle grid consists of repeated Rows
}

message Puzzle {
  repeated uint32 column_clues = 1 [packed=true];  // Packed encoding for repeated uint32 clues
  Grid grid = 2;                                  // The actual Puzzle grid
}

message PuzzleFile {
  uint32 number_of_puzzles = 1;    // Use uint32 (variable-length encoded)
  repeated Puzzle puzzles = 2;
}